contribs/gnodev/pkg/dev/node_test.go

```diff
@@ -0,0 +1,306 @@
+package dev
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/gnolang/gno/contribs/gnodev/pkg/emitter"
+	"github.com/gnolang/gno/contribs/gnodev/pkg/events"
+	"github.com/gnolang/gno/gno.land/pkg/gnoclient"
+	"github.com/gnolang/gno/gno.land/pkg/integration"
+	"github.com/gnolang/gno/gnovm/pkg/gnoenv"
+	core_types "github.com/gnolang/gno/tm2/pkg/bft/rpc/core/types"
+	"github.com/gnolang/gno/tm2/pkg/crypto/keys"
+	"github.com/gnolang/gno/tm2/pkg/log"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// XXX: We should probably use txtar to test this package
+
+// TestNewEmptyDevNode tests the NewDevNode method with no package
+func TestNewNode_NoPackages(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	logger := log.NewTestingLogger(t)
+
+	// Call NewDevNode with no package should works
+	cfg := DefaultNodeConfig(gnoenv.RootDir())
+	node, err := NewDevNode(ctx, logger, &emitter.NoopServer{}, cfg)
+	require.NoError(t, err)
+
+	assert.Len(t, node.ListPkgs(), 0)
+
+	require.NoError(t, node.Close())
+}
+
+// TestNodeWithPackage tests the NewDevNode with a single package.
+func TestNewNode_WithPackage(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	const (
+		// foobar package
+		testGnoMod = "module gno.land/r/dev/foobar\n"
+		testFile   = `package foobar
+func Render(_ string) string { return "foo" }
+`
+	)
+
+	// Generate package
+	pkgpath := generateTestingPackage(t, "gno.mod", testGnoMod, "foobar.gno", testFile)
+	logger := log.NewTestingLogger(t)
+
+	// Call NewDevNode with no package should works
+	cfg := DefaultNodeConfig(gnoenv.RootDir())
+	cfg.PackagesPathList = []string{pkgpath}
+	node, err := NewDevNode(ctx, logger, &emitter.NoopServer{}, cfg)
+	require.NoError(t, err)
+	assert.Len(t, node.ListPkgs(), 1)
+
+	// Test rendering
+	render, err := testingRenderRealm(t, node, "gno.land/r/dev/foobar")
+	require.NoError(t, err)
+	assert.Equal(t, render, "foo")
+
+	require.NoError(t, node.Close())
+}
+
+func TestNodeAddPackage(t *testing.T) {
```

Comment:
You can run this test in parallel

Reply:
duplicate of https://github.com/gnolang/gno/pull/1668#discussion_r1579881620